sdk/java/pulumi/src/test/java/com/pulumi/core/internal/ImportExportMetadataTest.java

```diff
@@ -20,22 +20,22 @@ class ImportExportMetadataTest {
     @SuppressWarnings("unused")
     public static class Tester {
 
-        @Export(name = "complex1", type = Either.class, parameters = {Integer.class, String.class})
+        @Export(name = "complex1", typeString = "com.pulumi.core.Either<java.lang.Integer,java.lang.String>")
         public final Output<Either<Integer, String>> complex1 = Output.of(Either.ofLeft(1));
 
-        @Export(name = "complex2", type = Either.class, parameters = {Integer.class, String.class})
+        @Export(name = "complex2", typeString = "com.pulumi.core.Either<java.lang.Integer,java.lang.String>")
```

Comment:
Hm. 

This looks workable, but it's worth thinking this through since this goes into providers and will be a long-term compatibility commitment. 

How about we ditch string representation and instead use `String.class` style references as before?

Using JSON notation so we do not need our own tree parser:

```
     Either<Map<String, List<String>>, Map<String, List<String>>>
     @Export(type="[0,[1,2,[3,2]],[1,2,[3,2]]]", refs=[Either.class, Map.class, String.class, List.class]

     Either<Integer, String>
     @Export(type="[0,1,2]", refs=[Either.class, String.class, Integer.class])
```

If prefer a bit more readable representation, go ahead and use our own tree parser:

```
     Either<Integer, String>
     @Export(type="0<1,2>", refs=[Either.class, String.class, Integer.class])
```

The more important aspect to me is if we could try and get to write `Either.class` instead of "com.pulumi.core.Either". This reuses JVM metadata handling and we don't need to invent class cache etc.

Whether we have a parser for our own format or no is less important - just need to review+test the parser thoroughy :)

Reply:
OK, this proposal looks more optimized and also more complex.

I'd like to kindly remind you @t0yv0 the string based representation was your idea.

Reply:
IMHO this boils down to where we move the complexity:
- string based moves the complexity to the SDK
- JSON+.class moves the complexity to the codegen

I like the idea of using .class, because we'd retain compile time check vs runtime lookup of classes based on strings.
On the other hand I like that with string based the modifications to the codegen are minimal.

Reply:
Decided on a sync to go ahead and move complexity to SDK 🙏 ping me again when ready to review (no rush).